### ISSUE TITLE
ENAMETOOLONG when uploading to Storage Emulator

### ISSUE THREAD

**Issue by starleaf1** [firebase/firebase-tools#4470]
### [REQUIRED] Environment info

**firebase-tools:**
10.7.1

**Platform:**
Ubuntu 20.04.4

### [REQUIRED] Test case

I have a React app set to run on Firebase Storage Emulator. I tried uploading to Storage and an error occurred. Debug log reveals that it's an `ENAMETOOLONG` error.

### [REQUIRED] Steps to reproduce

  1. Create a Firebase web project
  2. Set up Storage emulator and hook-up the app
  3. Upload a file with `uploadBytes` into Storage Emulator

### [REQUIRED] Expected behavior

File gets uploaded and shows up in Emulator UI.

### [REQUIRED] Actual behavior

Firebase emulator crashes, showing message `An unexpected error has occurred.` in console

[firebase-debug.log](https://github.com/firebase/firebase-tools/files/8559162/firebase-debug.log)

**Comment by joehan** [comment]
@tohhsinpei Could you take a look at this issue? I see that you've handled some Storage Emulator issues in the past. If you're the wrong person, could you reasssign?

**Comment by starleaf1** [comment]
I included the wrong log
[firebase-debug.log](https://github.com/firebase/firebase-tools/files/8559159/firebase-debug.log)
.

**Comment by justmorenoise** [comment]
Same issue on macOs 12.3.1
Firebase 9.6.6
Firebase Tools 10.8.0

```
(node:93295) UnhandledPromiseRejectionWarning: Error: ENAMETOOLONG: name too long, open '/var/folders/s4/xdtcj1yx4n34g1l_s5nlzp9m0000gn/T/firebase/storage/blobs/70f39d01-140e-456f-aac4-8cde0738144a_b_comicsflow.appspot.com_o_user_projects%252F2UpPaUVOEtHeawJAGyAEOJUI6qlY%252Fprojects%252Fab71b588-df19-4d8d-8684-c8ba612b38c8%252Fpages%252F2258b808-0f92-4471-a264-84b3ceb37e1f%252Factivities%252Fb9e9bb92-c008-4b63-9696-07dfcda952e7'
    at Object.openSync (fs.js:498:3)
    at Object.writeFileSync (fs.js:1524:35)
    at Object.appendFileSync (fs.js:1563:6)
    at Persistence.appendBytes (/usr/local/lib/node_modules/firebase-tools/lib/emulator/storage/persistence.js:25:16)
    at UploadService.multipartUpload (/usr/local/lib/node_modules/firebase-tools/lib/emulator/storage/upload.js:34:27)
    at handleUpload (/usr/local/lib/node_modules/firebase-tools/lib/emulator/storage/apis/firebase.js:184:42)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

The structure of the path is `user_projects/UUID/projects/UUID/pages/UUID/activities/UUID`

Works fine on cloud.

**Comment by IdanCo** [comment]
Same issue with the latest Firebase tools 11.0.0
macOS 12.3

**Comment by tohhsinpei** [comment]
Thanks all for the reports. I've managed to reproduce the issue; will try to put in a fix either this or next week.

**Comment by alexstanbury** [comment]
I'm also seeing this issue. My hunch is it's because somewhere in the process the url is being encoded, in my case the upload path had a lot of `%2F` in, which I suspect Node is interpreting to be one long string and causing this error.

Interestingly, I tried updating my firebase-tools, functions and node installation, and now I get a rather useless `Error: An unexpected error has occurred.` 

This seems to be affecting a fair few users, for example:
https://github.com/firebase/firebase-tools/issues/4588

After getting the useless error with the latest version I then tested by using a short file path and do not have any issues. In the issue linked above they are also using what looks like long paths.
